core/utils/STree.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Optional
+
+
+@dataclass
+class Node:
+    s_link: Optional[Node] = None
+    children: dict[str, Node] = field(default_factory=dict)
+    start: int = -1
+    depth: int = -1
+    parent: Optional[Node] = None
+
+    def __str__(self) -> str:
+        return f"children={list(self.children.keys())}"
+
+
+def create_node(s: str, u: Node, depth: int) -> Node:
+    start = u.start
+    parent = u.parent
+    v = Node()
+    v.start = start
+    v.depth = depth
+
+    v.children[s[start + depth]] = u
+    u.parent = v
+
+    parent.children[s[start + parent.depth]] = v
+    v.parent = parent
+
+    return v
+
+
+def create_and_attach_leaf(s: str, i: int, u: Node, depth: int) -> None:
+    leaf = Node()
+    leaf.start = i
+    leaf.depth = len(s) - i
+
+    u.children[s[i + depth]] = leaf
+    leaf.parent = u
+
+
+def compute_s_link(s: str, u: Node) -> None:
+    d = u.depth
+    v = u.parent.s_link
+
+    while v.depth < d - 1:
+        v = v.children[s[u.start + v.depth + 1]]
+
+    # if no node at (v, d - 1), then create one
+    if v.depth > d - 1:
+        v = create_node(s, v, d - 1)
```

Comment:
I see you allow nodes with out-degree 1 here. That is not a problem, of course, because you will fix it later; you only create it here when you will fix it later. You have a proof for that, I am sure.